pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java

```diff
@@ -239,11 +238,12 @@ public void removeSegment(IndexSegment segment) {
   }
 
   public static final class RecordInfo {
-    private final PrimaryKey _primaryKey;
+    /** stores the primary key of the record or its hash value */
+    private final Object _primaryKey;
     private final int _docId;
     private final Comparable _comparisonValue;
 
-    public RecordInfo(PrimaryKey primaryKey, int docId, Comparable comparisonValue) {
+    public RecordInfo(Object primaryKey, int docId, Comparable comparisonValue) {
```

Comment:
We should do the hash within the metadata manager instead of on the caller side. The primary key handling logic should all be wrapped within this class

Reply:
I thought about it. but this way makes the intermediate structure `RecordInfo` even lighter

Reply:
`RecordInfo` should wrap the necessary information, and it's up to the manager to process the info instead of the caller. It doesn't make `RecordInfo` lighter because both of them are just one reference

Reply:
hmm, to check we are speaking about the same thing.

do you mean
```
class RecordInfo {
    private final Object _primaryKey;
    private final int _docId;
    private final Comparable _comparisonValue;
   private final HashFunction _function;
```
by carrying additional field?

Reply:
No, `HashFunction` should be a member variable of `PartitionUpsertMetadataManager` which can be passed in from the constructor

Reply:
I see. That also works